File: MyApplication15/app/src/main/java/com/example/myapplication/MyView.java
package com.example.myview;

import android.content.Context;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.util.AttributeSet;
import android.view.View;

import androidx.annotation.Nullable;

import java.util.Collection;

public class MyView extends View {
    Paint mPaint;
    public MyView(Context context) {
        super( context );
        init();
    }

    public MyView(Context context, @Nullable AttributeSet attrs) {
        super( context, attrs );
        init();
    }

    public MyView(Context context, @Nullable AttributeSet attrs, int defStyleAttr) {
        super( context, attrs, defStyleAttr );
        init();
    }

    private void init()
    {
        mPaint=new Paint(Color.TRANSPARENT);
    }

    @Override
    protected void onMeasure(int widthMeasureSpec, int heightMeasureSpec) {
        super.onMeasure( widthMeasureSpec, heightMeasureSpec );
    }

    @Override
    protected void onLayout(boolean changed, int left, int top, int right, int bottom) {
        super.onLayout( changed, left, top, right, bottom );
    }

    @Override
    protected void onDraw(Canvas canvas) {
        super.onDraw( canvas );
        int size=Math.max(getWidth(),getHeight());
        canvas.drawCircle(getWidth()/2,getHeight()/2,size/2,mPaint);
    }
}

File: MyView6/app/src/main/java/com/example/myview/MyView.kt
package com.example.myview

import android.content.Context
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import android.util.AttributeSet
import android.view.View

class MyView(context: Context?, attrs: AttributeSet?, defStyleAttr: Int) :
    View(context, attrs, defStyleAttr) {

    val size=200

    val paint:Paint=Paint().apply {
        isAntiAlias=true
        style=Paint.Style.FILL
    }

    override fun onMeasure(widthMeasureSpec: Int, heightMeasureSpec: Int) {
        super.onMeasure(widthMeasureSpec, heightMeasureSpec)
        val widthSize=MeasureSpec.getSize(widthMeasureSpec)
        val widthMode=MeasureSpec.getMode(widthMeasureSpec)
        val heightSize=MeasureSpec.getSize(heightMeasureSpec)
        val heightMode=MeasureSpec.getMode(heightMeasureSpec)
        if(widthMode==MeasureSpec.AT_MOST&&heightMode==MeasureSpec.AT_MOST)
        {
            setMeasuredDimension(size,size)
        }
        else if(widthMode==MeasureSpec.AT_MOST)
        {
            setMeasuredDimension(size,heightSize)
        }
        else  if(heightMode==MeasureSpec.AT_MOST)
        {
            setMeasuredDimension(widthSize,size)
        }
    }

    override fun onLayout(changed: Boolean, left: Int, top: Int, right: Int, bottom: Int) {
        super.onLayout(changed, left, top, right, bottom)
        
    }

    override fun onDraw(canvas: Canvas?) {
        super.onDraw(canvas)
        val width=width.toFloat()
        val height=height.toFloat()
        val side= minOf(width,height)
        val left=width-side
        val top=height-side
        paint.color=Color.BLUE
        canvas?.drawRect(left,top,left+side,width+side,paint)
        val xx=left+side/2
        val yy=top+side/2
        paint.color=Color.RED
        canvas?.drawCircle(xx,yy,side/2,paint)
    }
}